Multi session support

anothrNick opened the pull request:
Added methods to support multiple sessions. Session.py keeps a dictionary of any open sessions. User can call `set_active_session` with a username to switch the active session and set the class variables (username, customer, password). The `log_off` method will log all open sessions out of the Dynect platform.

moogar0880 commented:
As it currently stands, this PR introduces breaking changes, and I don't believe it needs to. It probably makes more sense to create a new `SessionEngine`, or maybe even `DynectSession` subclass for handling multiple active sessions. Personally, I think that approach would make more sense as you're trying to hit functionality that the current `DynectSession` implementation was never intended to cover.

As we've seen before these changes to `DynectSession` will end up breaking existing code, and they don't need to

anothrNick commented:
Could you be more specific as to what this will break so I have a better understanding of how to design this feature? It was my understanding that the previous PR would break existing code since class variables were removed and there was an issue with the encryption of passwords. With those cleared up, it appears to be good.

I can absolutely create a subclass of `DynectSession` for this feature. But without knowing what broke I'm not quite sure how to proceed.

Thanks for reviewing!

anothrNick commented:
Or are you suggesting a subclass so the multi session feature must be used in a different way to avoid any unforeseen issues?

moogar0880 commented:
Specifically I'm concerned about the changes to the `log_out` method. 

But I think that, in general, I'm more concerned about your latter point. And not even nessecarily _unforseen_ issues, but the fundamental change in behavoir. Anyone using this module expecting to be able to only ever have a single active session likely won't be pleased if, all the sudden, they have hundreds of active sessions. 

If we put this into a new session type existing users can continue using this module without change, and whatever new users that want to run multiple active sessions in a single process can. 

And, to address your question about what broke last time, I'll let @Marchowes speak more to the point if he'd like, but from what I was told it was this change in behavior that caused accounts to get locked because they were hammering the api when, previously, they had just persisted a single session

anothrNick commented:
Thanks @moogar0880, it's a good point. I'll move the multi session functionality into a new subclass so we don't disrupt users who have no need for multiple sessions.

I had already discussed the issues from the last PR with Marc.

Thanks!

anothrNick commented:
I moved everything to `DynectMultiSession` which is a subclass of `DynectSession`. The only thing I changed in `DynectSession` was the addition of a `_encrypt` method so that the password can be encrypted in `DynectMultiSession` using the same AESCipher.

Marchowes commented on dyn/tm/session.py:
```diff
@@ -175,4 +181,89 @@ def __str__(self):
+        else:
+            raise ValueError("No open sessions for {0}".format(username))
+
+    def new_user_session(self, customer, username, password):
```
We shouldn't really require customer be included should we? I can see how switching session between customers could become troublesome... Are we expecting end users to use this to flip flop between different customer accounts?
